File: idea-plugin/src/main/java/com/zzf/rikki/idea/agent/LiteSseWriter.java
package com.zzf.rikki.idea.agent;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.zzf.rikki.idea.agent.compat.AgentEventSink;
import com.zzf.rikki.idea.agent.compat.RuntimeEvent;

import java.io.OutputStream;
import java.io.OutputStreamWriter;
import java.io.PrintWriter;
import java.nio.charset.StandardCharsets;
import java.util.LinkedHashMap;
import java.util.Map;

public class LiteSseWriter implements AgentEventSink {
    private final PrintWriter writer;
    private final ObjectMapper mapper = new ObjectMapper();

    public LiteSseWriter(OutputStream outputStream) {
        this.writer = new PrintWriter(new OutputStreamWriter(outputStream, StandardCharsets.UTF_8));
    }

    public synchronized void emit(String event, Object data) {
        writer.print("event: " + event + "\n");
        try {
            writer.print("data: " + (data instanceof String ? data : mapper.writeValueAsString(data)) + "\n");
        } catch (Exception e) {
            writer.print("data: {}\n");
        }
        writer.print("\n");
        writer.flush();
    }

    @Override
    public void emit(RuntimeEvent event) {
        if (event instanceof RuntimeEvent.SessionBound sessionBound) {
            emitSession(sessionBound.getSessionId(), sessionBound.getReused());
        } else if (event instanceof RuntimeEvent.StatusChanged statusChanged) {
            emitStatus(statusChanged.getType(), statusChanged.getMessage());
        } else if (event instanceof RuntimeEvent.MessageDelta messageDelta) {
            emitMessage(messageDelta.getMessageId(), messageDelta.getDelta());
        } else if (event instanceof RuntimeEvent.MessageSnapshot messageSnapshot) {
            emit("message_part", Map.of("messageID", messageSnapshot.getMessageId(), "text", messageSnapshot.getText()));
        } else if (event instanceof RuntimeEvent.ThoughtDelta thoughtDelta) {
            emitThought(thoughtDelta.getMessageId(), thoughtDelta.getDelta());
        } else if (event instanceof RuntimeEvent.ThoughtEnd thoughtEnd) {
            emitThoughtEnd(thoughtEnd.getMessageId());
        } else if (event instanceof RuntimeEvent.ToolCall toolCall) {
            emitToolCall(toolCall.getPartId(), toolCall.getTool(), toolCall.getCallId(), toolCall.getMessageId(), toolCall.getState(), toolCall.getTitle(), toolCall.getArgs(), toolCall.getMeta());
        } else if (event instanceof RuntimeEvent.ToolPendingApproval toolPendingApproval) {
            emitToolConfirm(toolPendingApproval.getPartId(), toolPendingApproval.getCallId(), toolPendingApproval.getCommand(), toolPendingApproval.getTool());
        } else if (event instanceof RuntimeEvent.ToolResult toolResult) {
            emitToolResult(toolResult.getPartId(), toolResult.getTool(), toolResult.getCallId(), toolResult.getMessageId(), toolResult.getState(), toolResult.getTitle(), toolResult.getOutput(), toolResult.getError(), toolResult.getMeta());
        } else if (event instanceof RuntimeEvent.TodoUpdated todoUpdated) {
            emitTodoUpdated(todoUpdated.getTodosJson(), todoUpdated.getSessionId());
        } else if (event instanceof RuntimeEvent.Finished finished) {
            emitFinish(finished.getSessionId(), finished.getMessageId(), finished.getAnswer());
        } else if (event instanceof RuntimeEvent.Errored errored) {
            emit("error", errored.getMessage());
        }
    }

    public void emitSession(String sessionId, boolean reused) {
        emit("session", Map.of("sessionID", sessionId, "reused", reused));
    }

    public void emitStatus(String type, String message) {
        emit("status", Map.of("type", type, "message", message));
    }

    public void emitMessage(String messageId, String delta) {
        emit("message", Map.of("id", messageId, "delta", delta));
    }

    public void emitThought(String messageId, String delta) {
        emit("thought", Map.of("messageID", messageId, "reasoning_delta", delta));
    }

    public void emitThoughtEnd(String messageId) {
        emit("thought_end", Map.of("messageID", messageId));
    }

    public void emitToolCall(String partId, String tool, String callId, String messageId, String state, String title, Object args, Map<String, ?> meta) {
        LinkedHashMap<String, Object> payload = new LinkedHashMap<>();
        payload.put("id", partId);
        payload.put("partID", partId);
        payload.put("partId", partId);
        payload.put("tool", tool);
        payload.put("callID", callId);
        payload.put("messageID", messageId);
        payload.put("messageId", messageId);
        payload.put("state", state);
        payload.put("title", title);
        payload.put("args", args);
        if (meta != null) {
            payload.put("meta", meta);
        }
        emit("tool_call", payload);
    }

    public void emitToolResult(String partId, String tool, String callId, String messageId, String state, String title, String output, String error, Map<String, ?> meta) {
        LinkedHashMap<String, Object> payload = new LinkedHashMap<>();
        payload.put("id", partId);
        payload.put("partID", partId);
        payload.put("partId", partId);
        payload.put("tool", tool);
        payload.put("callID", callId);
        payload.put("messageID", messageId);
        payload.put("messageId", messageId);
        payload.put("state", state);
        payload.put("title", title);
        payload.put("output", output);
        if (error != null) {
            payload.put("error", error);
        }
        if (meta != null) {
            payload.put("meta", meta);
        }
        emit("tool_result", payload);
    }

    public void emitToolConfirm(String partId, String callId, String command, String tool) {
        emit("tool_confirm", Map.of(
                "id", partId,
                "partID", partId,
                "partId", partId,
                "callID", callId,
                "callId", callId,
                "tool", tool,
                "command", command
        ));
    }

    public void emitTodoUpdated(String todosJson, String sessionId) {
        try {
            Object todos = mapper.readTree(todosJson);
            LinkedHashMap<String, Object> payload = new LinkedHashMap<>();
            payload.put("todos", todos);
            if (sessionId != null && !sessionId.isBlank()) {
                payload.put("sessionID", sessionId);
            }
            emit("todo_updated", payload);
        } catch (Exception ignored) {
        }
    }

    public void emitFinish(String sessionId, String messageId, String answer) {
        emit("finish", Map.of(
                "answer", answer,
                "traceId", sessionId,
                "sessionID", sessionId,
                "messageID", messageId,
                "meta", Map.of("pendingChanges", java.util.List.of())
        ));
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/agent/LiteSseWriter.kt
package com.zzf.rikki.idea.agent

import com.fasterxml.jackson.databind.ObjectMapper
import com.zzf.rikki.idea.agent.compat.AgentEventSink
import com.zzf.rikki.idea.agent.compat.RuntimeEvent
import java.io.OutputStream
import java.io.OutputStreamWriter
import java.io.PrintWriter
import java.nio.charset.StandardCharsets

/** Writes SSE events in the format ChatPanel/ChatSseAdapter expects. */
class LiteSseWriter(outputStream: OutputStream) : AgentEventSink {
    private val writer = PrintWriter(OutputStreamWriter(outputStream, StandardCharsets.UTF_8))
    private val mapper = ObjectMapper()

    @Synchronized
    fun emit(event: String, data: Any) {
        writer.print("event: $event\n")
        if (data is String) {
            writer.print("data: $data\n")
        } else {
            val json = mapper.writeValueAsString(data)
            writer.print("data: $json\n")
        }
        writer.print("\n")
        writer.flush()
    }

    override fun emit(event: RuntimeEvent) {
        when (event) {
            is RuntimeEvent.SessionBound -> emitSession(event.sessionId, event.reused)
            is RuntimeEvent.StatusChanged -> emitStatus(event.type, event.message)
            is RuntimeEvent.MessageDelta -> emitMessage(event.messageId, event.delta)
            is RuntimeEvent.MessageSnapshot -> emit("message_part", mapOf("messageID" to event.messageId, "text" to event.text))
            is RuntimeEvent.ThoughtDelta -> emitThought(event.messageId, event.delta)
            is RuntimeEvent.ThoughtEnd -> emitThoughtEnd(event.messageId)
            is RuntimeEvent.ToolCall -> emitToolCall(
                partId = event.partId,
                tool = event.tool,
                callId = event.callId,
                msgId = event.messageId,
                state = event.state,
                title = event.title,
                args = event.args,
                meta = event.meta
            )
            is RuntimeEvent.ToolPendingApproval -> emitToolConfirm(
                partId = event.partId,
                callId = event.callId,
                command = event.command,
                tool = event.tool
            )
            is RuntimeEvent.ToolResult -> emitToolResult(
                partId = event.partId,
                tool = event.tool,
                callId = event.callId,
                msgId = event.messageId,
                state = event.state,
                title = event.title,
                output = event.output,
                error = event.error,
                meta = event.meta
            )
            is RuntimeEvent.TodoUpdated -> emitTodoUpdated(event.todosJson, event.sessionId)
            is RuntimeEvent.Finished -> emitFinish(event.sessionId, event.messageId, event.answer)
            is RuntimeEvent.Errored -> emit("error", event.message)
        }
    }

    fun emitSession(sessionId: String, reused: Boolean = false) =
        emit("session", mapOf("sessionID" to sessionId, "reused" to reused))

    fun emitStatus(type: String, message: String) =
        emit("status", mapOf("type" to type, "message" to message))

    fun emitMessage(msgId: String, delta: String) =
        emit("message", mapOf("id" to msgId, "delta" to delta))

    fun emitThought(msgId: String, delta: String) =
        emit("thought", mapOf("messageID" to msgId, "reasoning_delta" to delta))

    fun emitThoughtEnd(msgId: String) =
        emit("thought_end", mapOf("messageID" to msgId))

    fun emitToolCall(
        partId: String,
        tool: String,
        callId: String,
        msgId: String,
        state: String,
        title: String,
        args: Any,
        meta: Map<String, Any?>? = null
    ) {
        val payload = linkedMapOf<String, Any?>(
            "id" to partId,
            "partID" to partId,
            "partId" to partId,
            "tool" to tool,
            "callID" to callId,
            "messageID" to msgId,
            "messageId" to msgId,
            "state" to state,
            "title" to title,
            "args" to args
        )
        if (meta != null) {
            payload["meta"] = meta
        }
        emit("tool_call", payload)
    }

    fun emitToolResult(
        partId: String,
        tool: String,
        callId: String,
        msgId: String,
        state: String,
        title: String,
        output: String,
        error: String? = null,
        meta: Map<String, Any?>? = null
    ) {
        val data = linkedMapOf<String, Any?>(
            "id" to partId,
            "partID" to partId,
            "partId" to partId,
            "tool" to tool,
            "callID" to callId,
            "messageID" to msgId,
            "messageId" to msgId,
            "state" to state,
            "title" to title,
            "output" to output
        )
        if (error != null) data["error"] = error
        if (meta != null) data["meta"] = meta
        emit("tool_result", data)
    }

    fun emitToolConfirm(partId: String, callId: String, command: String, tool: String) =
        emit(
            "tool_confirm", mapOf(
                "id" to partId,
                "partID" to partId,
                "partId" to partId,
                "callID" to callId,
                "callId" to callId,
                "tool" to tool,
                "command" to command
            )
        )

    fun emitTodoUpdated(todosJson: String, sessionId: String? = null) {
        try {
            val todosNode = mapper.readTree(todosJson)
            val payload = mutableMapOf<String, Any>("todos" to todosNode)
            if (!sessionId.isNullOrBlank()) {
                payload["sessionID"] = sessionId
            }
            emit("todo_updated", payload)
        } catch (_: Exception) {
        }
    }

    fun emitFinish(sessionId: String, msgId: String, answer: String) =
        emit(
            "finish", mapOf(
                "answer" to answer,
                "traceId" to sessionId,
                "sessionID" to sessionId,
                "messageID" to msgId,
                "meta" to mapOf("pendingChanges" to emptyList<Any>())
            )
        )
}